src/validation.cpp

```diff
@@ -399,6 +399,40 @@ void UpdateMempoolForReorg(DisconnectedBlockTransactions &disconnectpool, bool f
     LimitMempoolSize(mempool, GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000, GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60);
 }
 
+// Used to avoid mempool polluting consensus critical paths if CCoinsViewMempool
+// were somehow broken and returning the wrong scriptPubKeys
+static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, CValidationState &state, const CCoinsViewCache &view, CTxMemPool& pool,
+                 unsigned int flags, bool cacheSigStore, PrecomputedTransactionData& txdata) {
+    AssertLockHeld(cs_main);
+
+    assert(!tx.IsCoinBase());
+    {
+        LOCK(pool.cs);
```

Comment:
I'm still unconvinced it is needed, as it does not operate on CCoinsViewMemPool directly, but on a dissociated cache with the mempool UTXO loaded into it. If it's valid at one point with those inputs, it's also valid later.